cloudinit/sources/DataSourceAzure.py

```diff
@@ -208,6 +207,33 @@ def get_hv_netvsc_macs_normalized() -> List[str]:
     ]
 
 
+@azure_ds_telemetry_reporter
+def determine_device_driver_for_mac(mac: str) -> Optional[str]:
+    """Determine the device driver to match on, if any."""
+    drivers = [
+        i[2]
+        for i in net.get_interfaces(blacklist_drivers=BLACKLIST_DRIVERS)
+        if mac == normalize_mac_address(i[1])
```

Comment:
@cjp256 I don't think we need to call normalize_mac_address here on the return value from net.get_interfaces() right?
The output comes from /sys/class/net/<device>/address which should already be lowercase colon-delimited format.
```bash
$ python3 -c 'from cloudinit.net import get_interfaces; print(get_interfaces())'
[('lo', '00:00:00:00:00:00', None, None), ('eth0', '00:22:48:a6:0a:5c', 'hv_netvsc', '0x3')]
```


```suggestion
        if mac == i[1]
```

Reply:
As you mentioned a query in BSD over IRC. I poked around at a BSD 13.1 instance, and making the call to get_interfaces() may not have BSD support.
```bash
freebsd@bsd-131:~/cloud-init $ /usr/local/bin/python3 -c 'from cloudinit.net import get_interfaces; print(get_interfaces())'
[]
freebsd@bsd-131:~/cloud-init $ /usr/local/bin/python3 -c 'from cloudinit.net import get_interfaces; print(get_interfaces())'
```

Checking alternatives

Reply:
I agree get_interfaces() does not work on BSD but in this case it doesn't matter much.

(a) the blacklist_drivers used are very specific to Linux
(b) get_interfaces_by_mac() is already broken on BSD for Azure accelerated networking The mac of the accelerated interfaces clobbers the "master" interface, so networking never comes up properly.

```
>>> net.get_devicelist()
['mce0']
>>> net.get_interfaces_by_mac()
{'00:0d:3a:54:ad:1e': 'mce0'}
>>>
[root@test-freebsd131-x1 /usr/home/cpatterson]# ifconfig -a
lo0: flags=8049<UP,LOOPBACK,RUNNING,MULTICAST> metric 0 mtu 16384
        options=680003<RXCSUM,TXCSUM,LINKSTATE,RXCSUM_IPV6,TXCSUM_IPV6>
        inet6 ::1 prefixlen 128
        inet6 fe80::1%lo0 prefixlen 64 scopeid 0x1
        inet 127.0.0.1 netmask 0xff000000
        groups: lo
        nd6 options=21<PERFORMNUD,AUTO_LINKLOCAL>
hn0: flags=8843<UP,BROADCAST,RUNNING,SIMPLEX,MULTICAST> metric 0 mtu 1500
        options=8051b<RXCSUM,TXCSUM,VLAN_MTU,VLAN_HWTAGGING,TSO4,LRO,LINKSTATE>
        ether 00:0d:3a:54:ad:1e
        inet 10.0.0.35 netmask 0xffffff00 broadcast 10.0.0.255
        media: Ethernet 100GBase-CR4 <full-duplex,rxpause,txpause>
        status: active
        nd6 options=29<PERFORMNUD,IFDISABLED,AUTO_LINKLOCAL>
mce0: flags=8a43<UP,BROADCAST,RUNNING,ALLMULTI,SIMPLEX,MULTICAST> metric 0 mtu 1500
        options=8805bb<RXCSUM,TXCSUM,VLAN_MTU,VLAN_HWTAGGING,JUMBO_MTU,VLAN_HWCSUM,TSO4,LRO,LINKSTATE>
        ether 00:0d:3a:54:ad:1e
        media: Ethernet 100GBase-CR4 <full-duplex,rxpause,txpause>
        status: active
        nd6 options=29<PERFORMNUD,IFDISABLED,AUTO_LINKLOCAL>
```

I intend to purge blacklist_drivers from net/ and constrain this logic to Azure's data source.  Beyond that, to get it right, I think we'll need the BSD networking refactor to be done?

Reply:
Thoughts @igalic on the BSD angle?

Reply:
@cjp256 I've been working on getting blacklist driver support into BSD… 
but
are you saying that the blacklisted drivers are only relevant on Linux and their equivalent works on BSDs?

i'm also working on better get_interfaces and get_interfaces_by_mac into BSD, since obviously, that first shot using dictionaries wasn't great

the code you're demoing here isn't mine tho, and doesn't work much better

mine can be found at distros.bsd.networking.get_interfaces_by_mac()